Added a new feature to always check signature of already downloaded files

morettoni opened the pull request:
In our projects we use the maven-download-plugin, but sometimes other developers changes the content of the files to download without changing the file name.
This patch allow the developer to put a new option in the pom file (checkSignature) to check the signature the already downloaded files, if don't match the file is removed and downloaded again.

Hope this could be merged in the main code... let me know if I have it done in the correct way... is my first maven plugin!!!

mickaelistria commented:
Thanks for this report and patch.
I believe some related changes were tried previously. Before I look deeper in this patch, can you please confirm that the other story you've described was using the latest 1.2.1 release of download-maven-plugin?

morettoni commented:
Yes, I used 1.2.1, probably in the pom file you will find a wrong snapshot tag (could be I set it to 1.2.3...)

--
Luca Morettoni 

> On 12 Feb 2016, at 12:23, Mickael Istria <notifications@github.com> wrote:
> 
> Thanks for this report and patch.
> I believe some related changes were tried previously. Before I look deeper in this patch, can you please confirm that the other story you've described was using the latest 1.2.1 release of download-maven-plugin?
> 
> —
> Reply to this email directly or view it on GitHub.
>

mickaelistria commented on download-maven-plugin/pom.xml:
```diff
@@ -5,7 +5,7 @@
 	<artifactId>download-maven-plugin</artifactId>
 	<packaging>maven-plugin</packaging>
-	<version>1.2.2-SNAPSHOT</version>
+	<version>1.2.3-SNAPSHOT</version>
```
No need to change that. The 1.2.2 isn't release yet, it's our "work-in-progress version"

morettoni replied:
done

mickaelistria commented on download-maven-plugin/src/main/java/com/googlecode/WGet.java:
```diff
@@ -145,26 +145,32 @@
+   * Whether to check the signature of existing files
+   * @parameter expression="${checkSignature}" default-value="false"
+   */
+  private boolean checkSignature;
```
Actually, I'm wondering whether this should even be a parameter. Don't you think the check you suggest should be always performed?

morettoni replied:
About this change, usually when I'm adding a new feature to an existing package, the "default" value of what I have added is always what we had until now, so the check disabled.
But I don't see any problem to have it always active...

mickaelistria commented on download-maven-plugin/src/main/java/com/googlecode/WGet.java:
```diff
@@ -215,12 +221,46 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+
         // TODO verify last modification date
-        getLog().info("File already exist, skipping");
+        if (!haveFile)
```
Please always use { ... } blocks even for single instruction after a control statement. It prevents for future bugs with people adding instructions and not thinking about adding { ... } as well (that happened to me more than once ;)

morettoni replied:
done

mickaelistria commented on download-maven-plugin/src/main/java/com/googlecode/WGet.java:
```diff
@@ -215,12 +221,46 @@ public void execute() throws MojoExecutionException, MojoFailureException {
+            SignatureUtils.verifySignature(outputFile, this.md5,
+                    MessageDigest.getInstance("MD5"));
+          } catch (MojoFailureException e) {
+            getLog().warn("File already exist. " + e.getMessage());
```
This message doesn't seem very relevant here. I would suggest "The local cached version of file outputFile.getName() doesn't match the expected signature. You should consider checking the specified signature is correctly set, and/or cleaning your local cache."

morettoni replied:
done, here I also improved the code removing some dups and using a more clear (boolean) variable to perform the checks.

mickaelistria commented:
Except a few details commented inline, it seems to be a good change. Do you think you could manage to add an automated test for that user-story/bug?

morettoni commented:
I will implement/change your requests as soon as possible!

--
Luca Morettoni 

> On 13 Feb 2016, at 09:31, Mickael Istria <notifications@github.com> wrote:
> 
> Except a few details commented inline, it seems to be a good change. Do you think you could manage to add an automated test for that user-story/bug?
> 
> —
> Reply to this email directly or view it on GitHub.
>

morettoni commented:
Hi @mickaelistria sorry for the delay, I just added tests for my changes... I added some hacks to allow the creation of an empty file to force the usage of my option.
Let me know if I need to change other parts of the code!

mickaelistria commented:
Hi @morettoni . Sorry for the delay as well, I'm back from a week of holidays.
I've just looked quickly, and didn't spot anything bad.
Can you please squash your commits into a single one and use "push --force" on your master branch to update this PR to have a single commit? (Don't forget to push a backup on another branch before that ;)). Just so you know for future contribution, this project prefers single commits on PRs, so the best way to contribute is to use "git commit --amend" to fix your 1st commit and then use "git push --force" to update the PR.

mickaelistria commented:
Thanks. With this commit, when I run `mvn clean verify -Pits`, I get the following error:
```
[INFO] Building: basicGetWrongSignature/pom.xml
[INFO] ..FAILED (3.1 s)
[INFO]   The build exited with code 0. See /home/mistria/git/maven-download-plugin/download-maven-plugin/target/it/basicGetWrongSignature/build.log for details.
```
Looking inside the log, I see
```
[WARNING] Could not get content
org.apache.maven.plugin.MojoFailureException: Not same digest as expected: expected <wrong_md5> was <8412f316daecd4e8374a91fe13d9d657>
  at com.googlecode.SignatureUtils.verifySignature(SignatureUtils.java:35)
  at com.googlecode.WGet.execute(WGet.java:241)
  at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
  at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
  at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
  at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
  at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
  at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
  at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
  at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
  at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
  at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
  at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
  at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
  at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
  at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:497)
  at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
  at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
  at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
  at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
[WARNING] Not failing download despite download failure.
```
I believe your change actually changes the behavior for this test. Can you please have a look and recommend what's the best approach, keeping backward compatibility? It might be simply the test that should be updated since there are now more checks performed.

morettoni commented:
> On 23 Feb 2016, at 17:05, Mickael Istria <notifications@github.com> wrote:
> 
> [INFO] Building: basicGetWrongSignature/pom.xml
> [INFO] ..FAILED (3.1 s)
> [INFO]   The build exited with code 0. See /home/mistria/git/maven-download-plugin/download-maven-plugin/target/it/basicGetWrongSignature/build.log for details.
> Looking inside the log, I see
> 
> [WARNING] Could not get content
> org.apache.maven.plugin.MojoFailureException: Not same digest as expected: expected <wrong_md5> was <8412f316daecd4e8374a91fe13d9d657>
>   at com.googlecode.SignatureUtils.verifySignature(SignatureUtils.java:35)
>   at com.googlecode.WGet.execute(WGet.java:241)
Looks like there is a problem before my changes, I executed the tests over your code and I get the same:

INFO] Building: basicGetWrongSignature/pom.xml
[INFO] ..FAILED (4.7 s)
[INFO]   The build exited with code 0. See /Users/luca/src/maven-download-plugin-original/download-maven-plugin/target/it/basicGetWrongSignature/build.log for details.

and looking the log file the error is the same:

[WARNING] Could not get content
org.apache.maven.plugin.MojoFailureException: Not same digest as expected: expected <wrong_md5> was <8412f316daecd4e8374a91fe13d9d657>
        at com.googlecode.SignatureUtils.verifySignature(SignatureUtils.java:35)
        at com.googlecode.WGet.execute(WGet.java:292)

Tomorrow if you like I could try to fix it anyway, but I think we need to do that fix outside my patch, do you agree?

--
Luca Morettoni <luca(AT)morettoni.net> | http://www.morettoni.net
http://it.linkedin.com/in/morettoni/   | http://twitter.com/morettoni

morettoni commented:
> On 23 Feb 2016, at 17:05, Mickael Istria <notifications@github.com> wrote:
> 
> Thanks. With this commit, when I run mvn clean verify -Pits, I get the following error:
> 
> [INFO] Building: basicGetWrongSignature/pom.xml
> [INFO] ..FAILED (3.1 s)
> [INFO]   The build exited with code 0. See /home/mistria/git/maven-download-plugin/download-maven-plugin/target/it/basicGetWrongSignature/build.log for details.
About my previous comment, I think you need to add this line:

<failOnError>true</failOnError>

into src/it/basicGetWrongSignature/pom.xml file, otherwise maven will never fail even if the MD5 is wrong… :)

--
Luca Morettoni <luca(AT)morettoni.net> | http://www.morettoni.net
http://it.linkedin.com/in/morettoni/   | http://twitter.com/morettoni

mickaelistria commented:
Right, the issue doesn't seem related to your patch, so let's consider it in another report/PR.

mickaelistria commented:
Suggested commit was rebased and pushed as c5355e37f63ae3ebe8bd429af989c2fc70bdb136 .
A new 1.2.2-SNAPSHOT including that change has been deployed and can be consumed from OSSRH Maven Central.

Thanks!
